Benchmarks do not compile

```
$ cargo bench --all-features
   Compiling merkle_light_derive v0.3.1 (/Users/dignifiedquire/work/filecoin/merkle_light/merkle_derive)
   Compiling merkle_light v0.3.1 (/Users/dignifiedquire/work/filecoin/merkle_light/merkle)
warning: the item `std` is imported redundantly
  --> merkle/benches/ringx/init.rs:19:13
   |
19 |         use std;
   |             ^^^
   |
   = note: #[warn(unused_imports)] on by default

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:125:15
    |
125 |     b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/ring_sha512.rs:104:15
    |
104 |     b.iter(|| MerkleTree::<Hash512, B>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:125:15
    |
125 |     b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:131:15
    |
131 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:131:15
    |
131 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:131:40
    |
131 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:131:40
    |
131 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:144:15
    |
144 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:144:15
    |
144 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:144:40
    |
144 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:144:40
    |
144 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:159:15
    |
159 |     b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:159:15
    |
159 |     b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:165:15
    |
165 |     b.iter(|| MerkleTree::<Hash512, A>::from_par_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:165:15
    |
165 |     b.iter(|| MerkleTree::<Hash512, A>::from_par_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:171:15
    |
171 |     b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:171:15
    |
171 |     b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:177:15
    |
177 |     b.iter(|| MerkleTree::<Hash512, A>::from_par_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:177:15
    |
177 |     b.iter(|| MerkleTree::<Hash512, A>::from_par_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:183:15
    |
183 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:183:15
    |
183 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:183:40
    |
183 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:183:40
    |
183 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/crypto_sha512.rs:196:15
    |
196 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:196:15
    |
196 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:196:40
    |
196 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/crypto_sha512.rs:196:40
    |
196 |     let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:104:15
    |
104 |     b.iter(|| MerkleTree::<Hash512, B>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/ring_sha512.rs:110:15
    |
110 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:110:15
    |
110 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:110:40
    |
110 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:110:40
    |
110 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/ring_sha512.rs:123:15
    |
123 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:123:15
    |
123 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:123:40
    |
123 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:123:40
    |
123 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/ring_sha512.rs:138:15
    |
138 |     b.iter(|| MerkleTree::<Hash512, B>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:138:15
    |
138 |     b.iter(|| MerkleTree::<Hash512, B>::from_iter(values.clone()));
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/ring_sha512.rs:144:15
    |
144 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:144:15
    |
144 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:144:40
    |
144 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:144:40
    |
144 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0107]: wrong number of type arguments: expected 3, found 2
   --> merkle/benches/ring_sha512.rs:157:15
    |
157 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:157:15
    |
157 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |               ^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:157:40
    |
157 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error[E0277]: the trait bound `hash512::hash512::Hash512: merkle_light::merkle::Element` is not satisfied
   --> merkle/benches/ring_sha512.rs:157:40
    |
157 |     let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `merkle_light::merkle::Element` is not implemented for `hash512::hash512::Hash512`
    |
    = note: required by `merkle_light::merkle::MerkleTree`

error: aborting due to 26 previous errors

Some errors occurred: E0107, E0277.
For more information about an error, try `rustc --explain E0107`.
error: Could not compile `merkle_light`.
warning: build failed, waiting for other jobs to finish...
error: aborting due to 20 previous errors

Some errors occurred: E0107, E0277.
For more information about an error, try `rustc --explain E0107`.
error: Could not compile `merkle_light`.

To learn more, run the command again with --verbose.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -18,3 +18,4 @@ script:
   - cargo test --all-features
   - cargo doc --all-features --no-deps
   - bash -c 'if [[ "$TRAVIS_RUST_VERSION" == "nightly-2019-02-26" ]]; then cargo clippy --all -- -D warnings; fi'
+  - bash -c 'if [[ "$TRAVIS_RUST_VERSION" == "nightly-2019-02-26" ]]; then cargo build --benches --all-features --all --release; fi'

--- a/merkle/benches/crypto_sha512.rs
+++ b/merkle/benches/crypto_sha512.rs
@@ -14,7 +14,7 @@ use crypto::digest::Digest;
 use crypto::sha2::Sha512;
 use hash512::Hash512;
 use merkletree::hash::{Algorithm, Hashable};
-use merkletree::merkle::MerkleTree;
+use merkletree::merkle::{DiskMmapStore, MerkleTree, VecStore};
 use rand::Rng;
 use rayon::prelude::*;
 use std::hash::Hasher;
@@ -122,13 +122,13 @@ fn bench_crypto_sha512(b: &mut Bencher) {
 #[bench]
 fn bench_crypto_sha512_from_data_5(b: &mut Bencher) {
     let values = tree_5();
-    b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
+    b.iter(|| MerkleTree::<Hash512, A, VecStore<_>>::from_iter(values.clone()));
 }
 
 #[bench]
 fn bench_crypto_sha512_from_data_5_proof(b: &mut Bencher) {
     let values = tree_5();
-    let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
+    let tree: MerkleTree<Hash512, A, VecStore<_>> = MerkleTree::from_iter(values.clone());
 
     b.iter(|| {
         for i in 0..values.len() {
@@ -141,7 +141,7 @@ fn bench_crypto_sha512_from_data_5_proof(b: &mut Bencher) {
 #[bench]
 fn bench_crypto_sha512_from_data_5_proof_check(b: &mut Bencher) {
     let values = tree_5();
-    let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
+    let tree: MerkleTree<Hash512, A, VecStore<_>> = MerkleTree::from_iter(values.clone());
     let proofs = (0..values.len())
         .map(|i| tree.gen_proof(i))
         .collect::<Vec<_>>();
@@ -154,33 +154,45 @@ fn bench_crypto_sha512_from_data_5_proof_check(b: &mut Bencher) {
 }
 
 #[bench]
-fn bench_crypto_sha512_from_data_160(b: &mut Bencher) {
+fn bench_crypto_sha512_from_data_160_vec(b: &mut Bencher) {
     let values = tree_160();
-    b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
+    b.iter(|| MerkleTree::<Hash512, A, VecStore<_>>::from_iter(values.clone()));
+}
+
+#[bench]
+fn bench_crypto_sha512_from_data_160_mmap(b: &mut Bencher) {
+    let values = tree_160();
+    b.iter(|| MerkleTree::<Hash512, A, DiskMmapStore<_>>::from_iter(values.clone()));
 }
 
 #[bench]
 fn bench_crypto_sha512_from_data_160_par(b: &mut Bencher) {
     let values = tree_160();
-    b.iter(|| MerkleTree::<Hash512, A>::from_par_iter(values.clone()));
+    b.iter(|| MerkleTree::<Hash512, A, VecStore<_>>::from_par_iter(values.clone()));
 }
 
 #[bench]
-fn bench_crypto_sha512_from_data_30000(b: &mut Bencher) {
+fn bench_crypto_sha512_from_data_30000_vec(b: &mut Bencher) {
     let values = tree_30000();
-    b.iter(|| MerkleTree::<Hash512, A>::from_iter(values.clone()));
+    b.iter(|| MerkleTree::<Hash512, A, VecStore<_>>::from_iter(values.clone()));
+}
+
+#[bench]
+fn bench_crypto_sha512_from_data_30000_mmap(b: &mut Bencher) {
+    let values = tree_30000();
+    b.iter(|| MerkleTree::<Hash512, A, DiskMmapStore<_>>::from_iter(values.clone()));
 }
 
 #[bench]
 fn bench_crypto_sha512_from_data_30000_par(b: &mut Bencher) {
     let values = tree_30000();
-    b.iter(|| MerkleTree::<Hash512, A>::from_par_iter(values.clone()));
+    b.iter(|| MerkleTree::<Hash512, A, VecStore<_>>::from_par_iter(values.clone()));
 }
 
 #[bench]
 fn bench_crypto_sha512_from_data_160_proof(b: &mut Bencher) {
     let values = tree_160();
-    let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
+    let tree: MerkleTree<Hash512, A, VecStore<_>> = MerkleTree::from_iter(values.clone());
 
     b.iter(|| {
         for i in 0..values.len() {
@@ -193,7 +205,7 @@ fn bench_crypto_sha512_from_data_160_proof(b: &mut Bencher) {
 #[bench]
 fn bench_crypto_sha512_from_data_160_proof_check(b: &mut Bencher) {
     let values = tree_160();
-    let tree: MerkleTree<Hash512, A> = MerkleTree::from_iter(values.clone());
+    let tree: MerkleTree<Hash512, A, VecStore<_>> = MerkleTree::from_iter(values.clone());
     let proofs = (0..values.len())
         .map(|i| tree.gen_proof(i))
         .collect::<Vec<_>>();

--- a/merkle/benches/hash512/hash512.rs
+++ b/merkle/benches/hash512/hash512.rs
@@ -1,11 +1,37 @@
 use std::cmp::Ordering;
 
+use merkletree::merkle::Element;
+
 #[derive(Copy, Clone)]
 pub struct Hash512(pub [u8; 64]);
+
+impl std::fmt::Debug for Hash512 {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Hash512 {{ {:?} }}", &self.0[..])
+    }
+}
 
 impl Default for Hash512 {
     fn default() -> Self {
         Hash512([0u8; 64])
+    }
+}
+
+impl Element for Hash512 {
+    fn byte_len() -> usize {
+        64
+    }
+
+    fn from_slice(bytes: &[u8]) -> Self {
+        assert_eq!(bytes.len(), 64);
+        let mut a = Self::default();
+        a.0.copy_from_slice(bytes);
+
+        a
+    }
+
+    fn copy_to_slice(&self, bytes: &mut [u8]) {
+        bytes[..64].copy_from_slice(&self.0[..])
     }
 }
 

--- a/merkle/benches/ring_sha512.rs
+++ b/merkle/benches/ring_sha512.rs
@@ -11,7 +11,7 @@ extern crate test;
 
 use hash512::Hash512;
 use merkletree::hash::{Algorithm, Hashable};
-use merkletree::merkle::MerkleTree;
+use merkletree::merkle::{MerkleTree, VecStore};
 use rand::Rng;
 use ringx::Context;
 use ringx::SHA512;
@@ -101,13 +101,13 @@ fn bench_ringx_sha512(b: &mut Bencher) {
 #[bench]
 fn bench_ringx_sha512_from_data_5(b: &mut Bencher) {
     let values = tree_5();
-    b.iter(|| MerkleTree::<Hash512, B>::from_iter(values.clone()));
+    b.iter(|| MerkleTree::<Hash512, B, VecStore<_>>::from_iter(values.clone()));
 }
 
 #[bench]
 fn bench_ringx_sha512_from_data_5_proof(b: &mut Bencher) {
     let values = tree_5();
-    let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
+    let tree: MerkleTree<Hash512, B, VecStore<_>> = MerkleTree::from_iter(values.clone());
 
     b.iter(|| {
         for i in 0..values.len() {
@@ -120,7 +120,7 @@ fn bench_ringx_sha512_from_data_5_proof(b: &mut Bencher) {
 #[bench]
 fn bench_ringx_sha512_from_data_5_proof_check(b: &mut Bencher) {
     let values = tree_5();
-    let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
+    let tree: MerkleTree<Hash512, B, VecStore<_>> = MerkleTree::from_iter(values.clone());
     let proofs = (0..values.len())
         .map(|i| tree.gen_proof(i))
         .collect::<Vec<_>>();
@@ -135,13 +135,13 @@ fn bench_ringx_sha512_from_data_5_proof_check(b: &mut Bencher) {
 #[bench]
 fn bench_ringx_sha512_from_data_160(b: &mut Bencher) {
     let values = tree_160();
-    b.iter(|| MerkleTree::<Hash512, B>::from_iter(values.clone()));
+    b.iter(|| MerkleTree::<Hash512, B, VecStore<_>>::from_iter(values.clone()));
 }
 
 #[bench]
 fn bench_ringx_sha512_from_data_160_proof(b: &mut Bencher) {
     let values = tree_160();
-    let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
+    let tree: MerkleTree<Hash512, B, VecStore<_>> = MerkleTree::from_iter(values.clone());
 
     b.iter(|| {
         for i in 0..values.len() {
@@ -154,7 +154,7 @@ fn bench_ringx_sha512_from_data_160_proof(b: &mut Bencher) {
 #[bench]
 fn bench_ringx_sha512_from_data_160_proof_check(b: &mut Bencher) {
     let values = tree_160();
-    let tree: MerkleTree<Hash512, B> = MerkleTree::from_iter(values.clone());
+    let tree: MerkleTree<Hash512, B, VecStore<_>> = MerkleTree::from_iter(values.clone());
     let proofs = (0..values.len())
         .map(|i| tree.gen_proof(i))
         .collect::<Vec<_>>();

